### PR TITLE
Merge copybooks

### PR DESCRIPTION
**Pull request by tr11**
This PR adds functionality to merge distinct copybooks into a larger one.  It should address the second point in the discussion on https://github.com/AbsaOSS/cobrix/issues/33#issue-402115655, and make it indifferent for a user to provide multiple copybooks or a single aggregated one.

It's currently missing a corresponding option on the spark-cobol module.  Maybe we could allow the `copybook` option to be comma-separated paths? Or add a `copybooks` option to explicitly add more copybooks?

### CODE REVIEW

File: cobol-parser/src/test/scala/za/co/absa/cobrix/cobol/parser/copybooks/MergeCopybooksSpec.scala [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+ * limitations under the License.
+ */
+
+package za.co.absa.cobrix.cobol.parser.copybooks
+
+import org.scalatest.FunSuite
+import za.co.absa.cobrix.cobol.parser.{Copybook, CopybookParser}
+
+
+class MergeCopybooksSpec extends FunSuite {
```
Comment (yruslan) [inline cobol-parser/src/test/scala/za/co/absa/cobrix/cobol/parser/copybooks/MergeCopybooksSpec.scala]: Cheers for the good test coverage of the new feature!

File: cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/Copybook.scala [hunk 2]
```diff
@@ -236,5 +239,71 @@ class Copybook(val ast: CopybookAST) extends Serializable {
...
+        val head = cb.ast.children.head
+        if( !head.isRedefined || cb.ast.children.tail.exists(x => !x.redefines.contains(head.name)))
+          throw new RuntimeException("Copybook segments must redefine top segment.")
+      }
+    }
+
+    // create synthetic top level segment
+    val maxSize = copybooks.foldLeft(0) {(m, cb) => Math.max(m, cb.getRecordSize)}
+    val targetCopybook: Copybook = CopybookParser.parseTree(
+      s"      ${"%02d".format(rootLevel)} $targetName.\n       ${"%02d".format(rootLevel+1)} DATA PIC X($maxSize).\n"
```
[the remark this answers is not in the document]
Comment (tr11) [reply cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/Copybook.scala]: This was just to preserve the ordering of copybooks.  Not sure if that's too important, though.

We need to have the largest copybook first so the new groups redefine an area of the same size or possibly larger.  I opted to add an area that makes sure all the copybooks fit, but we could instead sort the copybooks by size and not add an extra area merge area if the input order is not important.

[the remark this answers is not in the document]
Comment (yruslan) [reply cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/Copybook.scala]: This is very interesting. The way redefines work in Cobrix is that the length of each redefined field is extended to the length of the biggest redefined field. So no matter what the order of the copybooks/segments is, the redefined fields should align properly and should have the same order as they were listed.

I'm wondering if the rule about the first redefined field needs to be the largest is related to COBOL spec? How redefines of different sizes behave for your use cases?

[the remark this answers is not in the document]
Comment (tr11) [reply cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/Copybook.scala]: I thought it was a spec requirement, but could be only the case for external data fields, according to https://www.ibm.com/support/knowledgecenter/en/SSUFAU_1.0.0/com.ibm.ent.cbl.zos.doc/PGandLR/ref/rlddered.html.

I'll remove that logic and add the `copybooks` option to the spark-cobol module.

[the remark this answers is not in the document]
Comment (yruslan) [reply cobol-parser/src/main/scala/za/co/absa/cobrix/cobol/parser/Copybook.scala]: Great, Thanks!

Took a look at the REDEFINED spec. Completely agree. It seems the rule applies only for external fields.

**Comment by tr11** [comment]
The tests are not too useful now besides showing that the feature works (i.e. two copybooks merged look like a single larger copybook).  We can extend the tests later to combine it with a custom record header.

**Comment by yruslan** [comment]
I like the feature and the tests. I think it is great!

Custom record headers is an orthogonal feature so it should work well in combination with merged copybooks. But yes, at some point in time it might be good to have a test that combines merging copybooks and custom record headers.

**Comment by yruslan** [comment]
Looked through the code again and it looks perfect. Merging...

- I'm going to take the feature to skip file header and trailing records to implement next week. I'm thinking offsets in bytes from the beginning and at the end of a file representing such records should be provided as an option. Cobrix will then skip that data while reading.

- I'd like to play some more with the ANTLR parser before merging. Considering it a part of a major feature release, like 0.6.0.

**Pull request merged**